### PR TITLE
Add support for handling url escaped path segments 

### PR DESCRIPTION
**Pull request by shafreenAnfar**
In url path segments following character are allowed, 

segment = *pchar
pchar = unreserved / pct-encoded / sub-delims / ":" / "@"

For more details please have a look at uri spec. 

If we need to have any other character than the aforementioned such as "?", "[", etc, we have to percent encode them. With this PR we support dispatching to such percent encoded path segments.

In addition to that I have removed all the unnecessary expressions that was there in URI Dispatching logic.

### CODE REVIEW

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 1]
```diff
@@ -521,31 +520,33 @@ public static void handleFailure(HTTPCarbonMessage requestMessage, BallerinaConn
...
+        }
+        setHttpStatusCodes(statusCode, response);
 
         return response;
     }
 
-    private static void setHttpStatusCodes(String payload, int statusCode, HTTPCarbonMessage response) {
+    private static String lowerCaseTheFirstLetter(String payload) {
+        char[] characters = payload.toCharArray();
+        characters[0] = Character.toLowerCase(characters[0]);
```
Comment (isudana) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Are we sure length of payload is not 0 here ?

Comment (shafreenAnfar) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Well, this payload is taken from BallerinaConnectorException and when you create BallerinaConnectorException we must give an informative error message as it is a common exception. But it was not enforced by the its constructors. hence, I update that class. Therefore, theoretically there shouldn't be an empty string.  

Anyways, just to be on the safe-side. I have checked for an empty string as well.

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java [hunk 2]
```diff
@@ -135,25 +156,35 @@ public static HttpResource findResource(HTTPServicesRegistry servicesRegistry,
...
-        Map<String, String> resourceArgumentValues =
-                (Map<String, String>) httpCarbonMessage.getProperty(Constants.RESOURCE_ARGS);
-
         BValue[] bValues = new BValue[paramDetails.size()];
         bValues[0] = request;
         bValues[1] = response;
         if (paramDetails.size() <= 2) {
             return bValues;
         }
+
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Can paramDetails.size() < 2 ? According to the above check it seems it is possible. If the size is < 2, the assignment for bValues[0] or bValues[1] may result in IndexOutofBoundsException .

Comment (shafreenAnfar) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Well, it cannot be less that two. It must to be more two or more.

**Pull request merged**